ArmPkg/Library/ArmMmuLib/AArch64/ArmMmuLibPageTableAlloc.c

```diff
@@ -0,0 +1,121 @@
+/** @file ArmMmuLibPageTableAlloc.c
+
+Logic facilitating the pre-allocation of page table memory.
+
+Copyright (C) Microsoft Corporation. All rights reserved.
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <Uefi.h>
+#include <Library/DebugLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/ArmMmuLib.h>
+
+typedef struct {
+  VOID     *NextPool;
+  UINTN    Offset;
+  UINTN    FreePages;
+} PAGE_TABLE_POOL;
+
+PAGE_TABLE_POOL  *mPageTablePool    = NULL;
+BOOLEAN          mPageTablePoolLock = FALSE;
+
+/**
+  Initializes the reserved pool used to allocate pages for page table memory.
+
+  @param[in] PoolPages The number of pages to initialize for the reserve pool. The actual number of pages
+                       reserved will be at least 512 - the number of pages required to describe a
+                       fully split 2MB region.
+
+  @retval EFI_SUCCESS             Reserved pool initialized successfully
+  @retval EFI_INVALID_PARAMETER   PoolPages is zero
+  @retval EFI_ABORTED             Software lock is held by another call to this function
+  @retval EFI_OUT_OF_RESOURCES    Unable to allocate pages
+**/
+EFI_STATUS
+InitializePageTablePool (
+  IN  UINTN  PoolPages
+  )
+{
+  VOID        *Buffer;
+  EFI_STATUS  Status = EFI_SUCCESS;
+
+  if (PoolPages == 0) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  //
+  // Avoid an infinite loop by bailing if we are nested in another call
+  // to this function
+  //
+  if (mPageTablePoolLock) {
+    return EFI_ABORTED;
+  }
+
+  mPageTablePoolLock = TRUE;
+  PoolPages++;
+  PoolPages = ALIGN_VALUE (PoolPages, EFI_SIZE_TO_PAGES (SIZE_2MB)); // Add one page for the header

```

Comment:
When we increment PoolPages we're adding an extra page, but we always allocate a size aligned to 2MB. So if a call is made to this to allocate 2MB, it will actually allocate 4MB and if a call is made to allocate 3MB it will also allocate 4MB.

Reply:
Oh, ha, I quite literally missed the increment directly above this.